Feature: Make remote position configurable

wutschel opened the pull request:
## Description
<!--- Detailed info for reviewers and developers -->
Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/343.
Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/332.

Feedback from users is that -- especially on bigger screens -- the remote is not usable by 1 hand anymore. This PR adds a configuration to select the position of the remote in the window. It is possible to select top, mid and bottom positions.

## Summary for release notes
<!--- This will be shown in Testflight to end users / testers -->
<!--- Please fill in, as usually PR title isn't clear to ordinary users -->
<!--- If your changes don't modify app files, please add prefix [not app] -->
Feature: Make remote position configurable
Feature: Add iPad's remote toolbar to iPhone
Feature: Add torch function to toolbar
Maintenance: Refactor torch handling to avoid duplicate code

wutschel commented:
Still marked as draft as I am not 100% satisfied with the safe area handling and the position calculation for newer iPhone.

wutschel commented:
With last push-force this works also for iPhone´s embedded view. Test on several devices to ensure the layout calculations are done right. 

Tested devices:
- iPodTouch 7th (iOS 14.5)
- iPhone 5s (iOS 11.4)
- iPhone 6s (iOS 11.4)
- iPhone Xs (iOS 14.5)
- iPad 5th (iOS 14.5)
- iPad Pro 7th 12.9" (iOS 14.5)

Open: Keep position after rotation on iPad.

wutschel commented:
Now rotation works.

Open: 
- Remove obsolete menu items from "user-defined-button" menu on iPhone (all functions are now available via remote screen toolbar)
- Remove non-used icons

wutschel commented:
Following screenshot shown the remote and custom button screen on iPhone with latest changes, and configured for showing the remote at bottom. I am thinking of also removing the "keyboard" line item as this is also not shown on the iPad custom screen. I kept it until now as the remote screen which is reachable via iPhone´s NowPlay screen is still showing this line item.

Screenshot: https://abload.de/img/bildschirmfoto2021-07qvk2l.png

Question: Are we losing any feature / use case when removing the line items like shown in the screen shot?

kambala-decapitator commented:
looks pretty neat in my opinion!

Does the new bottom bar look ok on the smallest device (iPhone 4S)?

> Question: Are we losing any feature / use case when removing the line items like shown in the screen shot?

you'd better ask the audience as I'm not a heavy remote user :) From a simple glance nothing gets lost.

@DakotaONeill what do you think?

wutschel commented:
> Does the new bottom bar look ok on the smallest device (iPhone 4S)?

I am not able to install the iPhone 4s simulator on my system. Seems it is requiring an older iOS as the oldest one which I can select in Xcode (iOS 11.4). But you right, I was testing with screen sizes with `height>=568`. Are you able to run a test?

wutschel commented:
> 2. Do not show the toolbar and keep supporting the functions/button on the custom button screen.

This was the easiest to implement. Last forced push has these changes.

In general I am not quite happy with the complex layout calculation and the special cases involved. I will try to look into supporting only top/bottom positions and use `UIViewAutoresizing` to align the remote view to top and bottom.

wutschel commented:
Huge rework to reduce the complexity. Now `UIViewAutoresizingFlexibleMargin` is used for the positioning and iPhone 4S should be supported properly. In consequence only the remote positions "top" and "bottom" are selectable.

wutschel commented:
Next bigger rework to also show a toolbar in the embedded remote view. As part of this rework
- a dedicated toolbar view is used which makes the layout easier
- the init-method has been updated to always initialize the remote view only once (before the layout was updated twice when entering the embedded view)
- the embedded only shows 4 buttons (as the 5th is not supported)

Screenshots (before and after):
https://abload.de/img/bildschirmfoto2021-07ljk96.png (iPhone Xs)
https://abload.de/img/bildschirmfoto2021-07p2je4.png (iPod Touch 7G)
https://abload.de/img/bildschirmfoto2021-07ehkzg.png (iPad 5G)

Remark 1: The "after" screenshots were made with the setting to move the remote screen to the bottom position. Of course it can still be positioned at the top -- still having the new toolbar.
Remark 2: For smaller devices (iPhone 4S) I cannot test. I took care of the small devices as far as I could, but it will be great if someone could provide screenshots.

wutschel commented:
It is even possible to add the volume slider to the remote screen for all screens with `height >= 568`. This looks a bit packed though: https://abload.de/img/simulatorscreenshot-icbjfn.png

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -270,6 +270,21 @@ + (LogoBackgroundType)getLogoBackgroundMode {
+        }
+        else if ([mode isEqualToString:@"position_bottom"]) {
+            setting = remoteBottom;
+        }
```
you could also add for the future:

```objc
else {
  NSAssert(NO, @"new mode not handled");
}
```

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -702,4 +703,51 @@ + (UIImage*)imageWithShadow:(UIImage*)source radius:(CGFloat)radius {
 }
 
++ (void)turnTorchOn:(id)sender on:(BOOL)torchOn {
+    Class captureDeviceClass = NSClassFromString(@"AVCaptureDevice");
+    if (captureDeviceClass != nil) {
```
this check is no longer needed

wutschel replied:
Removed

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -702,4 +703,51 @@ + (UIImage*)imageWithShadow:(UIImage*)source radius:(CGFloat)radius {
+        AVCaptureDevice *device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
+        AVCapturePhotoSettings *settings = [AVCapturePhotoSettings photoSettings];
+        UIImage *img;
+        if ([device hasTorch] && [device hasFlash]) {
```
please use early return: if not this condition then return

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -702,4 +703,51 @@ + (UIImage*)imageWithShadow:(UIImage*)source radius:(CGFloat)radius {
+            [device lockForConfiguration:nil];
+            // Set torch and select button image
+            if (torchOn) {
+                [device setTorchMode:AVCaptureTorchModeOn];
+                [settings setFlashMode:AVCaptureFlashModeOn];
```
please use properties

wutschel replied:
Done, also applied this for `hasTorch` and `hasFlash`.

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -702,4 +703,51 @@ + (UIImage*)imageWithShadow:(UIImage*)source radius:(CGFloat)radius {
+                [sender setImage:img forState:UIControlStateNormal];
+            }
+            else {
+                NSLog(@"Unexpected class. Cannot set image for torch icon.");
```
use `NSAssert` instead

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -702,4 +703,51 @@ + (UIImage*)imageWithShadow:(UIImage*)source radius:(CGFloat)radius {
+
++ (BOOL)isTorchOn {
+    BOOL torchIsOn = NO;
+    Class captureDeviceClass = NSClassFromString(@"AVCaptureDevice");
+    if (captureDeviceClass != nil) {
```
remove the check

wutschel replied:
Removed

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -702,4 +703,51 @@ + (UIImage*)imageWithShadow:(UIImage*)source radius:(CGFloat)radius {
+    }
+}
+
++ (BOOL)isTorchOn {
```
you could also return `nil` when device has no torch/flash, if that makes any sense for the calling code

wutschel replied:
I would like to keep it as it is, just returning plain `YES` or `NO`.

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -177,8 +177,8 @@ - (void)configureView {
-        frame.size.width = [[UIScreen mainScreen] bounds].size.width;
-        frame.origin.x = ((remoteControlView.frame.size.width - [[UIScreen mainScreen] bounds].size.width) / 2);
+        frame.size.width = UIScreen.mainScreen.bounds.size.width;
+        frame.origin.x = ((remoteControlView.frame.size.width - UIScreen.mainScreen.bounds.size.width) / 2);
```
please remove the outer brackets

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -45,6 +42,23 @@ - (void)setDetailItem:(id)newDetailItem {
+    gestureImage = [UIImage imageNamed:@"finger"];
+    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+    BOOL showGesture = [[userDefaults objectForKey:@"gesture_preference"] boolValue];
+    if (showGesture) {
```
early return

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -45,6 +42,23 @@ - (void)setDetailItem:(id)newDetailItem {
     }
 }
 
+- (void)setGestureView {
```
please come up with a better method name

wutschel replied:
`setupGestureView` ? This is what it does.

kambala-decapitator replied:
yep, that's better

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -111,43 +125,39 @@ - (void)setEmbeddedView {
-    [gestureButton setBackgroundImage:[UIImage imageNamed:@"remote_button_blank_down"] forState:UIControlStateHighlighted];
-    gestureButton.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleTopMargin;
-    [gestureButton addTarget:self action:@selector(toggleGestureZone:) forControlEvents:UIControlEventTouchUpInside];
+    [remoteControlView setFrame:CGRectMake(0, offset, newWidth, newHeight)];
```
use property

wutschel replied:
Done. I have 100+ calls of `setFrame` in the code. I can change this with a coding style commit.

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -1235,6 +1250,68 @@ - (void)turnTorchOn:(id)sender {
+    CGFloat ToolbarFlexSpace = ((width - 5 * TOOLBAR_ICON_SIZE) / 6);
+    CGFloat ToolbarPadding = (TOOLBAR_ICON_SIZE + ToolbarFlexSpace);
+    CGFloat ToolbarStart = 5 * ToolbarPadding;
+    CGRect frame = CGRectMake(CGRectGetMaxX(self.view.frame) - xMin - ToolbarStart, yMax - TOOLBAR_ICON_SIZE - TOOLBAR_FIXED_OFFSET, TOOLBAR_ICON_SIZE, TOOLBAR_ICON_SIZE);
```
are you sure that using `self.view.frame` is right? You're computing frames for subviews here.

oh, I see you removed this code in the next commit.

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -1235,6 +1250,68 @@ - (void)turnTorchOn:(id)sender {
     [Utilities turnTorchOn:sender on:torchIsOn];
 }
 
+- (void)createRemoteToolbar:(UIImage*)gestureButtonImg width:(CGFloat)width xMin:(CGFloat)xMin yMax:(CGFloat)yMax {
```
I'd rather use `UIStackView` to create a toolbar.

wutschel replied:
Is blocking to approve the PR? I can look into this later.

kambala-decapitator replied:
not at all

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -1235,6 +1250,68 @@ - (void)turnTorchOn:(id)sender {
+    CGRect frame = CGRectMake(CGRectGetMaxX(self.view.frame) - xMin - ToolbarStart, yMax - TOOLBAR_ICON_SIZE - TOOLBAR_FIXED_OFFSET, TOOLBAR_ICON_SIZE, TOOLBAR_ICON_SIZE);
+    UIButton *settingButton = [UIButton buttonWithType:UIButtonTypeCustom];
+    settingButton.frame = frame;
+    [settingButton setContentMode:UIViewContentModeRight];
```
please use properties where possible

wutschel replied:
This is already removed. But there are several calls of `setContentMode` which I can replace with using the property directly. Will do this in a separate commit.

kambala-decapitator reviewed (changes requested):
may I ask you to split 37d12f5c74065301ef17b334c02e9501dd707ad8 into multiple? It contains a lot of separate changes and is quite difficult to review.

kambala-decapitator commented:
> It is even possible to add the volume slider to the remote screen for all screens with `height >= 568`. This looks a bit packed though: https://abload.de/img/simulatorscreenshot-icbjfn.png

looks fine I'd say

wutschel commented:
> may I ask you to split [37d12f5](https://github.com/xbmc/Official-Kodi-Remote-iOS/commit/37d12f5c74065301ef17b334c02e9501dd707ad8) into multiple? It contains a lot of separate changes and is quite difficult to review.

I am sorry, and in general I would have loved to do this. Because of massive refactoring in the course of implementing this feature, I couldn´t find a good way to squash certain changes. That´s why this ended up in one huge commit. Also, I lost my old backup branch when I needed to recreate my local repository. Only option would be to re-implement or split out changes manually now.

How important is this for you?

kambala-decapitator commented:
ok, hoped it's straightforward :) will try to review as is.

wutschel commented:
> > It is even possible to add the volume slider to the remote screen for all screens with `height >= 568`. This looks a bit packed though: https://abload.de/img/simulatorscreenshot-icbjfn.png
> 
> looks fine I'd say

Ok, will see if I can add this on top of this PR after all the fixups were made and squashed.

kambala-decapitator reviewed (changes requested):
fixups are fine.

also reviewed that huge commit.

> Handle small iPhone displays (width < 576, e.g. iPhone 4S)

width or height?

wutschel commented:
> fixups are fine.

Good. Will do the other fixups and find a way to squash this.

> width or height?

height. Will fix the commit's description.

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -66,16 +73,18 @@ - (void)setEmbeddedView {
                 hide:YES];
-    if ([[UIScreen mainScreen] bounds].size.height >= 568) {
+    if ([self hasRemoteToolBar]) {
+        shift = ((UIButton*)[self.view viewWithTag:21]).frame.size.height;
```
Done

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -66,16 +73,18 @@ - (void)setEmbeddedView {
+                    ypos: -shift];
     }
     else {
+        shift = 2.0 * ((UIButton*)[self.view viewWithTag:21]).frame.size.height;
```
Done

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -92,7 +108,16 @@ - (void)setEmbeddedView {
+    else {
+        remoteControlView.autoresizingMask = UIViewAutoresizingFlexibleBottomMargin;
+    }
+    [remoteControlView setFrame:CGRectMake(-1, offset, newWidth, newHeight)];
```
Already done in later commit.

wutschel commented on XBMC Remote/RightMenuViewController.m:
```diff
@@ -186,19 +191,19 @@ - (UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSI
             [onoff setAutoresizingMask:icon.autoresizingMask];
             [onoff addTarget: self action: @selector(toggleSwitch:) forControlEvents:UIControlEventValueChanged];
-            [onoff setFrame:CGRectMake(0, cellHeight/2 - onoff.frame.size.height/2, onoff.frame.size.width, onoff.frame.size.height)];
+            [onoff setFrame:CGRectMake(0, (RIGHT_MENU_ITEM_HEIGHT - onoff.frame.size.height) / 2, onoff.frame.size.width, onoff.frame.size.height)];
```
Done

wutschel commented on XBMC Remote/RightMenuViewController.m:
```diff
@@ -261,12 +266,12 @@ - (UIView*)createTableFooterView:(CGFloat)footerHeight {
         frame.size.width = STACKSCROLL_WIDTH;
         [fixedSpace setWidth:0.0];
-        [toolbar setFrame:CGRectMake(0, 0, frame.size.width, 44)];
+        [toolbar setFrame:CGRectMake(0, 0, frame.size.width, TOOLBAR_HEIGHT)];
```
Done

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -129,44 +154,66 @@ - (void)configureView {
         CGRect frame = remoteControlView.frame;
-        frame.size.height = frame.size.height *transform;
-        frame.size.width = frame.size.width*transform;
+        frame.size.height = frame.size.height * transform;
+        frame.size.width = frame.size.width * transform;
```
Done

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -58,6 +64,7 @@ - (void)hideButton:(NSArray*)buttonsToDo hide:(BOOL)hide {
     CGFloat newWidth = remoteControlView.frame.size.width - ANCHOR_RIGHT_PEEK;
+    CGFloat shift;
     [self hideButton: [NSArray arrayWithObjects:
                        [(UIButton*)self.view viewWithTag:2],
```
You mean the tag IDs?

kambala-decapitator replied:
yep

wutschel replied:
Done. Found 1 obsolete case-statement as an outcome. Reads much better now.

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -129,44 +154,66 @@ - (void)configureView {
+            remoteControlView.autoresizingMask = UIViewAutoresizingFlexibleBottomMargin;
+        }
         remoteControlView.frame = frame;
+        frame.origin.y = 0;
+        quickHelpView.frame = frame;
```
Between the assignment of frame, to easier differentiate between different assignments? Like this:

```
remoteControlView.frame = frame;
        
frame.origin.y = 0;
quickHelpView.frame = frame;
        
frame = subsInfoLabel.frame;
```

kambala-decapitator replied:
exactly

wutschel replied:
Ok, will do after the other fixups are accepted and I could squash them.

kambala-decapitator replied:
sure no problem

kambala-decapitator replied:
sure no problem

wutschel replied:
Done

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -129,44 +154,66 @@ - (void)configureView {
-        CGRect frame = subsInfoLabel.frame;
-        frame.size.width = newWidth;
+        CGRect frame = remoteControlView.frame;
+        frame.size.height = CGRectGetHeight(remoteControlView.frame) * transform;
+        frame.size.width = CGRectGetWidth(remoteControlView.frame) * transform;
```
Good catch. I find the `*=` quite well readable. Moving this into a helper could be done, but might be less quickly readable.

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -1241,25 +1288,25 @@ - (void)viewDidLoad {
+    if ([self hasRemoteToolBar]) {
+        torchIsOn = NO;
+        Class captureDeviceClass = NSClassFromString(@"AVCaptureDevice");
+        if (captureDeviceClass != nil) {
```
Yes, already removed with one of the earlier fixup commits.

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -1273,7 +1320,7 @@ - (void)viewDidLoad {
         gestureButton.alpha = 0.8;
         [self.view addSubview:gestureButton];
-        
+            
```
Done

wutschel commented on XBMC Remote/RightMenuViewController.m:
```diff
@@ -156,7 +161,7 @@ - (UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSI
     }
     else {
-        int cellHeight = 50;
+
```
Already removed later.

wutschel commented on XBMC Remote/RightMenuViewController.m:
```diff
@@ -122,17 +128,16 @@ - (UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSI
         [icon setAutoresizingMask:UIViewAutoresizingNone];
-        [icon setFrame:CGRectMake(10, (int)((cellHeight/2) - (18/2)), 18, 18)];
-        [title setFrame:CGRectMake(icon.frame.size.width + 16, (int)((cellHeight/2) - (title.frame.size.height/2)), tableView.frame.size.width - (icon.frame.size.width + 32), title.frame.size.height)];
+        [icon setFrame:CGRectMake(10, (SERVER_INFO_HEIGHT - RIGHT_MENU_ICON_SIZE) / 2, RIGHT_MENU_ICON_SIZE, RIGHT_MENU_ICON_SIZE)];
+        [title setFrame:CGRectMake(icon.frame.size.width + RIGHT_MENU_ICON_SPACING, (SERVER_INFO_HEIGHT - title.frame.size.height) / 2, tableView.frame.size.width - (icon.frame.size.width + 2 * RIGHT_MENU_ICON_SPACING), title.frame.size.height)];
```
Done

wutschel reviewed (commented):
Test

wutschel commented on XBMC Remote/RightMenuViewController.m:
```diff
@@ -58,12 +64,12 @@ - (void)turnTorchOn:(BOOL)on icon:(UIImageView*)iconTorch {
     }
     else if ([tableData[indexPath.row][@"label"] isEqualToString:@"RemoteControl"]) {
-        return 570;
+        return [[UIScreen mainScreen] bounds].size.height - [self getRemoteViewOffsetY];
```
Why not simply `UIScreen.mainScreen.bounds.size.height` and `CGRectGetHeight` instead? A fixup commit already did this.

kambala-decapitator replied:
ok, let's leave like that

wutschel commented on XBMC Remote/RightMenuViewController.m:
```diff
@@ -276,30 +281,40 @@ - (UIView*)createTableFooterView:(CGFloat)footerHeight {
+    // Layout is (top-down): status bar > server info > volume slider > (menu items) > remote view
+    CGFloat statusBarHeight = [UIApplication sharedApplication].statusBarFrame.size.height;
+    CGFloat sliderHeight = volumeSliderView.frame.size.height;
+    CGFloat menuItemsHeight = UIScreen.mainScreen.bounds.size.height >= 568 ? 0 : 3 * RIGHT_MENU_ITEM_HEIGHT;
```
Yes. Should I create a helper function in `Utilities`?

kambala-decapitator replied:
yes please

wutschel replied:
Done

wutschel commented on XBMC Remote/RightMenuViewController.m:
```diff
@@ -785,17 +807,22 @@ - (void)setRightMenuOption:(NSString*)key reloadTableData:(BOOL)reload {
-                              @(NO), @"isSetting",
-                              @"embedded", @"type",
-                            nil]];
+        NSMutableDictionary *itemDict = [NSMutableDictionary dictionaryWithObjectsAndKeys:
```
I double checked this. Mutable is not required, I also changed to literal in a fixup commit.

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -92,7 +108,16 @@ - (void)setEmbeddedView {
-    [remoteControlView setFrame:CGRectMake(-1, 0, newWidth, newHeight)];
+    CGFloat offset = 0;
+    RemotePositionType positionMode = [Utilities getRemotePositionMode];
+    if (positionMode == remoteBottom && [self hasRemoteToolBar]) {
```
This is possible, let me do this once the other fixups are ok and I could squash them. Otherwise this get too nasty squash -- maybe it is already...

kambala-decapitator replied:
sure

wutschel replied:
Done

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -129,44 +154,66 @@ - (void)configureView {
+            remoteControlView.autoresizingMask = UIViewAutoresizingFlexibleBottomMargin;
+        }
+        remoteControlView.frame = frame;
+        frame.origin.y = 0;
+        frame.origin.x = 0;
+        quickHelpView.frame = frame;
```
CGPointZero change is done. The empty lines I will look at when the other fixups are accepted.

wutschel replied:
Empty lines also done.

wutschel commented on XBMC Remote/RemoteController.m:
```diff
@@ -84,6 +93,13 @@ - (void)setEmbeddedView {
     }
+    // Keep the buttons out of the safe area
+    CGFloat bottomPadding = 0.0;
+    if (@available(iOS 11.0, *)) {
```
Agreed, let me do this in another PR as this touches code all around. Will create an issue for this.

wutschel commented:
Most of the fixups and rework is added now. Only the requested empty lines and the possible refactoring of `if (positionMode == remoteBottom && [self hasRemoteToolBar]) {` is not done yet. I really need to see, if I can squash the fixups before I move one.

kambala-decapitator commented:
looks good, please proceed

wutschel commented:
Squashed all fixup commits -- except one change (the space correction stays separate) -- and rebased to master. Will now look into the last open review comments.

wutschel commented:
Added the refactoring around `(positionMode == remoteBottom && [self hasRemoteToolBar])` and added empty lines for the `= frame` occurances. If ok, I will try to squash especially the "empty line" changes into the older commits. If this is getting too difficult, I will just squash them into one single "empty lines" commit. I hope this is fine for you.

kambala-decapitator commented:
Yep, fine

wutschel commented:
I decided to just add the last fixups as two separate commits. Squashing them back into the former commits is too much work and not worth it.

kambala-decapitator commented:
feedback by @DakotaONeill 

> Hey OG poster who submitted request for movable screen. I’ll reply to your email at another time but wanted to send this here. Regardless of being at the top or bottom. The iPhone X and forward phones that have a virtual home button (the line at the bottom) is overlapping the toolbar you have shifted to the bottom. Everything else seems good though. Again will reply to email when I have a chance but figured this would be helpful in the meantime (on vacation).

<img src="https://user-images.githubusercontent.com/1557784/133071418-e147ac4f-2169-4319-bd00-7ee16f660349.jpg" width="20%" />

wutschel commented:
In fact, I did this for purpose. This way there is no space wasted and while testing I didn't find this problematic. As far as I remember I even found other Apps doing the same. But now I cannot find such anymore ... Change it, or leave it?

kambala-decapitator commented:
controls overlapping "safe area" is considered a bad UX, would be good to adjust

wutschel commented:
> controls overlapping "safe area" is considered a bad UX, would be good to adjust

Please see https://github.com/xbmc/Official-Kodi-Remote-iOS/pull/390
